src/dataframe/dataframe.jl

```diff
@@ -1070,6 +1074,14 @@ julia> df1
 ```
 """
 function Base.append!(df1::DataFrame, df2::AbstractDataFrame)
+    if ncol(df1) == 0
+        for (n, v) in eachcol(df2, true)
+            df1[n] = collect(v) # make sure df1 contains Vector-s
```

Comment:
This makes a copy even when not necessary.

Reply:
I thought the opposite. Given column "ownership" contract we always have to make a copy. When do you think copy could be dropped?

Reply:
Ah, sorry, of course here we create new column vectors so we need to make the copy here. But I don't really understand your comment in the OP about `AbstractVector`: why not just use the type of `v` (i.e. use `copy(v)`)?

Reply:
I was considering `copy`. The only downside is that for non standard `AbstractVectors` you do not know what you will get. E.g. if the source data frame contains `SVector`s then `copy` returns an `SVector` and `SVector` cannot be resized (and usually if you run `append!` you probably want to do it several times). `collect` guarantees we get a `Vector`.

But I agree that the choice here is not clear cut and we could use `copy` (and actually this is the problem with allowing `appending!` to an empty `DataFrame`).

Reply:
`collect` wouldn't give the right result for `CategoricalArray`. We could use `copyto!(Tables.allocatecolumn(eltype(v), length(v)), v)`, but it could lose information contained by the array. Since AFAIK DataFrames hasn't really been tested with static arrays, I'd go for the simpler `copy` for now.

Reply:
OK - fixed.